src/wp-includes/blocks.php

```diff
@@ -886,38 +886,68 @@ function serialize_blocks( $blocks ) {
+ * `$pre_callback` receives the previous block, whereas the `$post_callback` receives
+ * the next block as third argument.
+ *
+ * Serialized blocks are returned including comment delimiters, and with all attributes serialized.
+ *
+ * This function should be used when there is a need to modify the saved block, or to inject markup
+ * into the return value. Prefer `serialize_block` when preparing a block to be saved to post content.
  *
  * @since 6.4.0
  *
  * @see serialize_block()
  *
- * @param array    $block    A representative array of a single parsed block object. See WP_Block_Parser_Block.
- * @param callable $callback Callback to run on each block in the tree before serialization.
- *                           It is called with the following arguments: $block, $parent_block, $block_index, $chunk_index.
- * @return string String of rendered HTML.
+ * @param array    $block         A representative array of a single parsed block object. See WP_Block_Parser_Block.
+ * @param callable $pre_callback  Callback to run on each block in the tree before it is traversed and serialized.
+ *                                It is called with the following arguments: &$block, $parent_block, $previous_block.
+ *                                Its string return value will be prepended to the serialized block markup.
+ * @param callable $post_callback Callback to run on each block in the tree after it is traversed and serialized.
+ *                                It is called with the following arguments: &$block, $parent_block, $next_block.
+ *                                Its string return value will be appended to the serialized block markup.
+ * @return string Serialized block markup.
  */
-function traverse_and_serialize_block( $block, $callback ) {
+function traverse_and_serialize_block( $block, $pre_callback = null, $post_callback = null ) {
 	$block_content = '';
 	$block_index   = 0;
 
-	foreach ( $block['innerContent'] as $chunk_index => $chunk ) {
+	foreach ( $block['innerContent'] as $chunk ) {
 		if ( is_string( $chunk ) ) {
 			$block_content .= $chunk;
 		} else {
-			$inner_block = call_user_func(
-				$callback,
-				$block['innerBlocks'][ $block_index ],
-				$block,
-				$block_index,
-				$chunk_index
-			);
+			$inner_block = $block['innerBlocks'][ $block_index ];
+
+			if ( is_callable( $pre_callback ) ) {
+				$prev = 0 === $block_index
+					? null
+					: $block['innerBlocks'][ $block_index - 1 ];
+				$block_content .= call_user_func_array(
+					$pre_callback,
+					array( &$inner_block, $block, $prev )
+				);
+			}
+
+			$block_content .= traverse_and_serialize_block( $inner_block, $pre_callback, $post_callback );
+
+			if ( is_callable( $post_callback ) ) {
+				$next = count( $block['innerBlocks'] ) - 1 === $block_index
+					? null
+					: $block['innerBlocks'][ $block_index + 1 ];
+				$block_content .= call_user_func_array(
+					$post_callback,
+					array( &$inner_block, $block, $next )
+				);
+			}
```

Comment:
funny enough this kind of thing is where I've been exploring the use of a generator function as a context manager.

```php
foreach ( traverse_and_serialize_block( $original_block ) as &$block ) {
	// do anything here
}
```

this kind of structure in PHP gets around some of the more awkward properties of passing objects/arrays as function arguments, or of the need to explicitly `use` closed values.

guess it probably doesn't handle the recursion that well though.